### ISSUE TITLE
The script stopped working

### ISSUE THREAD

**Issue by icue** [icue/SteamWishlistCalendar#13]
As of now, steam wishlist can't be viewed by guests, even when it's set to public.

Not sure if it's a temporary bug or a decision made by Valve, in which case authentication would be needed for the script to read the wishlist.

The issue is echoed in some other reports: https://forum.gg.deals/d/1509-is-steam-wishlist-sync-broken-or-is-it-just-me

**Comment by icue** [comment]
At the very least, scheduled workflow should be disabled to prevent wishlist data from disappearing.

@Vinfall FYI.

**Comment by Vinfall** [comment]
I guess it's possibly related to a Steam backend incident recently. Someone accidentally rollbacked the database and I happened to notice that in the exact same moment (many old games, even delisted ones appear in wishlist again last week, but was fixed within a few minutes).

~~Further research led to [this guide](https://steamcommunity.com/sharedfiles/filedetails/?id=1746978201) on Steam Community which seems to validate my point. I think it also explains the misbehavior in https://github.com/icue/SteamWishlistCalendar/issues/11#issuecomment-2294186421, kind of... Maybe there are some hidden items previously.~~ Did not notice you mentioned it already.

gg.deals said they deployed a fix to the bot, but I've yet to see how SteamWishlistCalendar could do the same.

**Comment by Vinfall** [comment]
TLDR: `/wishlist/xxx/xxx/wishlistdata/` seems no longer publicly available (which is very weird so I doubt I missed something), the only *working* solution is web scraping and even that only works 60-80% of the time...



I roughly digged a bit, some findings (I'll omit `https://store.steampowered.com` url prefix in the following lines):
- `/wishlist/id/{args.id}/wishlistdata/` or `/wishlist/profiles/{args.id}/wishlistdata/` never worked for me
- If `id` is a full Steam ID (e.g. `1234567890123456`), it would not work (aka. `/wishlist/profiles/`)
- If `id` is community id (e.g. `gabenewell`), then
	- `/wishlist/id/{args.id}/` or `/wishlist/id/{args.id}/wishlistdata/` not working
	- `/wishlist/id/{args.id}` (note the trailing `/` is removed), has 80% chance to work…

I'm not very sure, but there seems no network request to Steam API in developer tools. Instead, the only request related to wishlist (when visiting `/wishlist/id/{args.id}` via browser) is a GET request that would return the whole HTML. This does not make sense, so I highly doubt if I missed something…

Good thing is the request headers has set a cookie containing (partial) wishlist data, but it's not enough for a wishlist calendar. I'm also not sure how things would change for wishlist with more items (for example, 2k+ games).

Example `app_impressions` cookie (there is also `wishlistPrefs` but I guess it's unrelated):

```json
[
    {
        "name": "app_impressions",
        "value": "1749810%401_24_5__777%7C620980%401_24_5__777%7C2616280%401_24_5__777",
        "domain": "store.steampowered.com",
        "hostOnly": true,
        "path": "/",
        "secure": true,
        "httpOnly": false,
        "sameSite": "no_restriction",
        "session": true,
        "firstPartyDomain": "",
        "partitionKey": null,
        "storeId": null
    }
]
```

One liner to parse it in jq:
- `1_24_5__777` is the tracking param `snr` commonly seen in Steam URL `?snr=1_24_5__777`
- `1749810,620980,2616280` in final result are AppId

```sh
# decode
$ jq -r '.[] | select(.name == "app_impressions") | .value' < 1.json | sed 's/%/\\x/g' | xargs -0 printf

1749810@1_24_5__777|620980@1_24_5__777|2616280@1_24_5__777

# remove snr between `@` and `|`
$ jq -r '.[] | select(.name == "app_impressions") | .value' < 1.json | sed 's/%/\\x/g' | xargs -0 printf | sed 's/@[^|]*//g' | sed 's/|/,/g'

1749810,620980,2616280
```

**Comment by icue** [comment]
Thanks a lot for the investigation, @Vinfall 
Actullay I've been kind of procrastinating the fix on my end because, knowing the development velocity of Valve, part of me just wish it's a bug on their end which they will _eventually_ fix and then everything will get back to normal.
But on the other hand, I know that's unlikely to happen.

Anyways, I noticed two websites that are still functioning:
* https://www.steamwishlistcalculator.com/
* https://www.lorenzostanco.com/lab/steam/wishlist/

As for the first one, it's open source, and [this commit](https://github.com/The-HopelessGamer/steamwishlistcalculator/commit/4501696068557930fb04aa15eab60e3f3632eae3) indicates that they basically had to rewrite the website due to the recent change from Valve. I have yet to read their codebase to understand how they manage to get user wishlist data.

I've also reached out to the author of the second website. Maybe I can get some useful insights.

Hopefully there's a solution that's both reliable and does not require a server.

**Comment by icue** [comment]
People are also discussing the problem here:
https://www.reddit.com/r/Steam/comments/1h1rkce/how_to_get_wishlist_data_from_steam_api/

**Comment by Vinfall** [comment]
So there is now an official [GetWishlist API](https://steamapi.xpaw.me/#IWishlistService/GetWishlist)? I swear it wasn't there back in [issue#11](https://github.com/icue/SteamWishlistCalendar/issues/11#issuecomment-2297850349).

<details><summary>Details</summary>
<p>

~~Anyway, it returns protobuf instead of JSON…~~ Bingo, you can get wishlist via `https://api.steampowered.com/IWishlistService/GetWishlist/v1/?steamid=1234567890123456` but it's similar to my example cookie above: only appid, no name or release date.

For appid/name mapping at least, you can query [GetAppList API](https://steamapi.xpaw.me/#ISteamApps/GetAppList) to have them all at once.
`jq '.applist.apps | length' 1.json` shows there are 224800+ items so it should be abundant.
The result would be something like this:

```json
{
  "applist": {
    "apps": [
      {"appid": 3341330, "name": "サモンズソード ～ 鍋の守護者 ～"},
      {"appid": 3341350, "name": "Crazy Voyage"},
      {"appid": 3341360, "name": "Heath Demo Playtest"}
    ]
  }
}
```


</p>
</details> 


For less elegant solution using these APIs, it's quite simple. You just would hit rate limit *very* quickly:
1. get wishlist appids via `GetWishlist` API
2. get name mapping via `GetAppList` API
3. get release date somehow
	- [SteamSpy API](https://steamspy.com/api.php) allows 1 request/second, but `appdetails` does not have release date (e.g. `https://steamspy.com/api.php?request=appdetails&appid=410340`), so it's not useful here.
	- From [a very old Reddit post](https://old.reddit.com/r/Steam/comments/bcmqjn/how_many_requests_to_steam_am_i_allowed_per/), JSON API allows 69 request/minute.
	- In the worst case, query one by one via `https://store.steampowered.com/api/appdetails?appids={APPID}`. BTW this way you can have `short_description` added to the calendar.
4. generate ics with these metadata

PS: Personally I'd rather use a userscript to generate calendar on demand instead. It just has to be manual but I only check game release once a week so it's fine for me.

**Comment by icue** [comment]
Thanks for the input.

Yeah it seems that the wishlist API is newly added. (As a result, maybe #11 has been be fixed by the new API. We'll see.)

Since the response of `https://store.steampowered.com/api/appdetails?appids={APPID}` already contains the name of the game, I think I can just use that to get name & release date together.

From [this reddit post](https://www.reddit.com/r/Steam/comments/1h1rkce/comment/lzixn2y/), there is also [GetItems](https://steamapi.xpaw.me/#IStoreBrowseService/GetItems) to get the details of a group of appids **in one API call**. However, it requires a steam API key, so it's harder to set up or distribute.

I'll go with the plain old `appdetails` approach for now (until Valve takes down that API, too, then I'll migrate to the API key approach).

**Comment by Vinfall** [comment]
I noticed the new commits, here are my suggestions.

With proper caching, requests can be reduced a lot:
- Querying `GetAppList` API to have AppId & their name mapped (to have a large mapping initialized before making request) in a database.[^1]
- Cache metadata of *released* games[^2] somewhere (e.g. a SQLite database), as it's unlikely to change (release date, description).
- Before requesting `appdetails`, query AppId in the database above.
	- If it does not exist in the database, make a new `GetAppList` call and update existing database
	- If it exists
		- Released: do not make new request. Instead, simply return the information.
		- Unreleased: request as usual.

For example:
- After winter sale, I only have 20+ items in wishlist. With caching, the number goes down to…just 9.
- You only need to query 64+2 (bad appid) out of 184 items with caching.


[^1]: Side note: I guess it's a misbehavior? At least this file should be excluded in the license.
[^2]: Another note: actually, I don't think it's even necessary to include very old releases (released 2 months or a year ago?), as you won't find them in calendar without searching…So this database can be used as an exclusion list in that case.

**Comment by icue** [comment]
Thanks for the suggestions.

I was actually trying to avoid cache because I want the script to be as stateless and "portable" as possible. (Admittedly the whole "wishlist count history tracking diagram" feature is pretty stateful - I'll probably split that into a different module in the future to keep the ICS generation module clean.)

The good news is that, I tried the [GetItems](https://steamapi.xpaw.me/#IStoreBrowseService/GetItems) API **without** specifying a steam API key, and it worked. It can reliably get the detailed information of something like 200 games at once. I tried raising that number to 300 and got an error response. So I batch process 200 games at a time.

Based on my experiments, repeated runs also produce the same result, and no game is missing. So this is a better method overall.

I'll close this issue. Let me know if it does not work on your end.

**Comment by Vinfall** [comment]
Glad that API works without dev key, let's see how long it would last...
